cli/npm/byonm.rs

```diff
@@ -288,29 +334,62 @@ impl CliNpmResolver for ByonmCliNpmResolver {
     req: &PackageReq,
     referrer: &ModuleSpecifier,
   ) -> Result<PathBuf, AnyError> {
-    // resolve the pkg json and alias
-    let (pkg_json, alias) =
-      self.resolve_pkg_json_and_alias_for_req(req, referrer)?;
-    // now try node resolution
-    for ancestor in pkg_json.path.parent().unwrap().ancestors() {
-      let node_modules_folder = ancestor.join("node_modules");
-      let sub_dir = join_package_name(&node_modules_folder, &alias);
-      if self.fs.is_dir_sync(&sub_dir) {
-        return Ok(canonicalize_path_maybe_not_exists_with_fs(
-          &sub_dir,
-          self.fs.as_ref(),
-        )?);
+    fn node_resolve_dir(
+      fs: &dyn FileSystem,
+      alias: &str,
+      start_dir: &Path,
+    ) -> Result<Option<PathBuf>, AnyError> {
+      for ancestor in start_dir.ancestors() {
+        let node_modules_folder = ancestor.join("node_modules");
+        let sub_dir = join_package_name(&node_modules_folder, alias);
+        if fs.is_dir_sync(&sub_dir) {
+          return Ok(Some(canonicalize_path_maybe_not_exists_with_fs(
+            &sub_dir, fs,
+          )?));
+        }
       }
+      Ok(None)
     }
 
-    bail!(
-      concat!(
-        "Could not find \"{}\" in a node_modules folder. ",
-        "Deno expects the node_modules/ directory to be up to date. ",
-        "Did you forget to run `deno install`?"
-      ),
-      alias,
-    );
+    // now attempt to resolve if it's found in any package.json
+    let maybe_pkg_json_and_alias =
+      self.resolve_pkg_json_and_alias_for_req(req, referrer)?;
+    match maybe_pkg_json_and_alias {
+      Some((pkg_json, alias)) => {
+        // now try node resolution
+        if let Some(resolved) =
+          node_resolve_dir(self.fs.as_ref(), &alias, pkg_json.dir_path())?
+        {
+          return Ok(resolved);
+        }
+
+        bail!(
+          concat!(
+            "Could not find \"{}\" in a node_modules folder. ",
+            "Deno expects the node_modules/ directory to be up to date. ",
+            "Did you forget to run `deno install`?"
+          ),
+          alias,
+        );
+      }
+      None => {
+        // now check if node_modules/.deno/ matches this constraint
+        if let Some(folder) = self.resolve_folder_in_root_node_modules(req) {
+          return Ok(folder);
+        }
+
+        bail!(
+          concat!(
+            "Could not find a matching package for 'npm:{}' in the node_modules ",
+            "directory. Ensure you have all your JSR and npm dependencies listed ",
+            "in your deno.json or package.json, then run `deno install`. Alternatively, ",
+            "turn on auto-install by specifying `\"nodeModulesDir\": \"auto\"` in your ",
```

Comment:
Not a big deal, but for easier grepping it would be better to have this line use `r#""#` so you don't have to escape quotes. TIOLI